Detect process.exit() called from tests

gibson042 opened the pull request:
Fixes #861


<!-- Issuehunt content -->

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Referenced issues

This pull request has been submitted to:
- [#861: Override `process.exit` to give some context when it's called unexpectedly in a test.](https://oss.issuehunt.io/repos/26820798/issues/861)
---
</details>
<!-- /Issuehunt content-->

gibson042 committed:
Override process.exit in tests

Fixes #861

gibson042 committed:
Add process.exit test

novemberborn commented on lib/worker/base.js:
```diff
@@ -19,6 +19,35 @@ import {flags, refs, sharedWorkerTeardowns} from './state.cjs';
+
+process.exit = new Proxy(realExit, {
+	apply(fn, receiver, args) {
+		channel.send({type: 'internal-error', err: serializeError('Forced exit error', false, new Error('Unexpected process.exit()'), runner.file)});
```
Could leave `runner` undefined and use `runner?.file`, that should work in the supported Node.js versions I believe.

gibson042 replied:
Done.

novemberborn commented on lib/worker/base.js:
```diff
@@ -19,6 +19,35 @@ import {flags, refs, sharedWorkerTeardowns} from './state.cjs';
+
+process.exit = new Proxy(realExit, {
+	apply(fn, receiver, args) {
+		channel.send({type: 'internal-error', err: serializeError('Forced exit error', false, new Error('Unexpected process.exit()'), runner.file)});
```
I think `internal` should refer to AVA doing something wrong, but this is triggered by user code. What do you think?

gibson042 replied:
I think that's valid, although there is also an argument to be made that early termination before all tests run is an internal error. But what should this be instead?

novemberborn replied:
`process-exit` maybe? Would also allow us to print a specific message, like we do for `SIGINT`:

https://github.com/avajs/ava/blob/392ab65cd4d203643e5bb29aa7f152056fb6040b/lib/reporters/default.js#L195-L200

gibson042 replied:
Done.

novemberborn reviewed (commented):
Amazing!

gibson042 committed:
Replace dummy initialization with optional chain

gibson042 committed:
Create a process-exit message for process.exit()

gibson042 committed:
Mitigate a racy test

A worker cannot observe a peer-failed message sent before it started listening.

novemberborn commented:
@gibson042 I should have some time this week to help get this over the line.

novemberborn committed:
Merge remote-tracking branch 'origin/main' into gh-861-process-exit

novemberborn committed:
Remove unused interrupts array from default reporter

novemberborn committed:
Specific output in TAP reporter

novemberborn committed:
Only include the stack to the actual process.exit() call

No need for other error details, the error is synthetic anyway.

novemberborn committed:
process-exit always comes from a test file

novemberborn committed:
Start flushing even when exiting synchronously

novemberborn committed:
Add ava.config.js to fixture so we can run AVA from the directory

novemberborn commented:
@gibson042 thanks for [`96ff6ef` (#3080)](https://github.com/avajs/ava/pull/3080/commits/96ff6ef0b80cba3d42d2237cb7e3d410ce9b8776)! Let's see if it improves things.

I've pushed several updates, please let me know what you think.

gibson042 committed:
Improve a variable name

gibson042 commented:
@novemberborn The updates look good to me. My only further change was to rename the generic `proxyApply` to `handleProcessExit`.
